web/packages/teleport/src/DesktopSession/useTdpClientCanvas.tsx

```diff
@@ -194,18 +241,34 @@ export default function useTdpClientCanvas(props: Props) {
    * The remote Windows machine also treats CapsLock like a normal key, and
    * expects a DOWN-UP whenever it's pressed.
    */
-  const handleCapsLock = (cli: TdpClient, e: KeyboardEvent): boolean => {
-    if (e.code === 'CapsLock' && isMac) {
-      cli.sendKeyboardInput(e.code, ButtonState.DOWN);
-      cli.sendKeyboardInput(e.code, ButtonState.UP);
+  const handleCapsLock = (
+    cli: TdpClient,
+    e: KeyboardEvent,
+    state: ButtonState
+  ): boolean => {
+    if (e.code === 'CapsLock') {
+      if (isMac) {
+        // On Mac, every UP or DOWN given to us by the browser corresponds
+        // to a DOWN + UP on the remote machine.
+        cli.sendKeyboardInput(e.code, ButtonState.DOWN);
+        cli.sendKeyboardInput(e.code, ButtonState.UP);
+      } else {
+        // On Windows or Linux, we just pass the event through normally to the server.
+        cli.sendKeyboardInput(e.code, state);
```

Comment:
Was this an unrelated bug? Where were we doing this before?

Reply:
We were doing this here

https://github.com/gravitational/teleport/blob/07514ea6deee86194aa55a1a52128fbd7a452a9b/web/packages/teleport/src/DesktopSession/useTdpClientCanvas.tsx#L171-L178

and then using it like

https://github.com/gravitational/teleport/blob/07514ea6deee86194aa55a1a52128fbd7a452a9b/web/packages/teleport/src/DesktopSession/useTdpClientCanvas.tsx#L180-L183

This change is functionally the same, I've just consolidated all of the special capslock logic in one place.

Reply:
I refactored it again here since it was still unnecessarily complicated, now the control flow is more straightforward.

42a1e76c5e5270b0779200b2d6384daff9b6da6b